lib/sample_rate_converter/sample_rate_converter.c

```diff
@@ -0,0 +1,357 @@
+#endif
+
+static uint8_t internal_input_buf[INTERNAL_INPUT_BUF_SIZE];
+static uint8_t internal_output_buf[INTERNAL_OUTPUT_BUF_SIZE];
+
+static int validate_sample_rates(uint32_t input_sample_rate, uint32_t output_sample_rate)
+{
+	if (input_sample_rate > output_sample_rate) {
+		if (input_sample_rate != 48000) {
+			LOG_ERR("Invalid input sample rate for downsampling %d", input_sample_rate);
+			return -EINVAL;
+		}
+
+		if ((output_sample_rate != 24000) && (output_sample_rate != 16000)) {
+			LOG_ERR("Invalid output sample rate for downsampling %d",
+				output_sample_rate);
+			return -EINVAL;
+		}
+	} else if (input_sample_rate < output_sample_rate) {
+		if (output_sample_rate != 48000) {
+			LOG_ERR("Invalid output sample rate for upsampling: %d",
+				output_sample_rate);
+			return -EINVAL;
+		}
+
+		if ((input_sample_rate != 24000) && (input_sample_rate != 16000)) {
+			LOG_ERR("Invalid input sample rate for upsampling: %d", input_sample_rate);
+			return -EINVAL;
+		}
+	} else {
+		LOG_ERR("Input and out samplerates are the same");
+		return -EINVAL;
+	}
+	return 0;
+}
+
+static bool is_upsampling(uint32_t input_rate, uint32_t output_rate)
+{
+	return input_rate < output_rate;
+}
+
+/**
+ * When upsampling from 16kHz to 48kHz the input and output bytes must be buffered. This is to
+ * fulfill the requirement for the filter that the number of input bytes must be divisible by the
+ * conversion factor. This function returns true when this is the case.
+ */
+static bool conversion_needs_buffering(struct sample_rate_converter_ctx *ctx,
+				       uint8_t conversion_ratio)
+{
+	return (is_upsampling(ctx->input_sample_rate, ctx->output_sample_rate) &&
+		(conversion_ratio == 3));
+}
+
+static uint8_t calculate_conversion_ratio(uint32_t input_sample_rate, uint32_t output_sample_rate)
+{
+	if (input_sample_rate > output_sample_rate) {
+		return input_sample_rate / output_sample_rate;
+	} else {
+		return output_sample_rate / input_sample_rate;
+	}
+}
+
+/**
+ * @brief Initializes the sample rate converter context
+ *
+ * @details Validates and sets all sample rate conversion parameters for the context. If buffering
+ *	    is needed for the conversion, the input buffer will be padded with two samples to
+ *	    ensure there will always be enough samples for a valid conversion.
+ *
+ * @param[in,out]	ctx			Pointer to the sample rate conversion context
+ * @param[in]		input_sample_rate	Sample rate of the input samples
+ * @param[in]		output_sample_rate	Sample rate of the output samples
+ * @param[in]		filter			Filter type to use in the conversion
+ *
+ * @retval 0 On success
+ * @retval -EINVAL Invalid parameters used to initialize the conversion
+ * @retval Other errno codes from CMSIS DSP library
+ */
+static int sample_rate_converter_init(struct sample_rate_converter_ctx *ctx,
+				      uint32_t input_sample_rate, uint32_t output_sample_rate,
+				      enum sample_rate_converter_filter filter)
```

Comment:
This is a really good point that I hadn't thought of.

As we've discussed offline I've renamed `sample_rate_converter_init` to `sample_rate_converter_reconfigure`, and I've added `sample_rate_converter_open` to 0-ize the context. I'm not sure if `_init` or `_open` is the best, but I went with `_open` as I feel `_init` is stricter - but I'm not married to the name if you think otherwise